### PR TITLE
Improve ``devices`` and related documentation

### LINKED ISSUE #2898: jax.devices returns only the devices from the default backend
The documentation of jax.devices says that it returns all devices. This is not true in multi-backend setups, it returns only the devices from the default backend. This was confusing to me initially, and I also encountered this confusion in issue #2785 

The simplest change would be to the documentation: if no `backend` is specified then only the devices on the default backend are returned (along with an explanation of what is the default backend). 

A better change may be though to say that it returns all devices, in the order of backend priority, such that `devices()[0]` is the same as now. This may break some code though.

### PR DESCRIPTION
**Pull request by skye**
Fixes #2898

### CODE REVIEW

File: jax/lib/xla_bridge.py [hunk 1]
```diff
@@ -202,14 +204,26 @@ def devices(backend=None):
   return get_backend(backend).devices()
 
 
-def local_devices(host_id=None, backend=None):
-  """Returns a list of devices local to a given host (this host by default)."""
+def local_devices(host_id: int = None, backend: str = None):
+  """Like ``devices``, but only returns devices local to a given host.
+
+  If ``host_id`` is ``None``, returns devices local to this host.
```
Comment (gnecula) [inline jax/lib/xla_bridge.py:210]: Will this give an error if we pick the host_id and a backend that does not exist on that host?

File: jax/lib/xla_bridge.py [hunk 2]
```diff
@@ -179,19 +179,21 @@ def device_count(backend=None):
...
 
-  Each device is represented by a subclass of Device (e.g. CpuDevice,
-  GpuDevice). The length of the returned list is equal to
-  ``device_count()``. Local devices can be identified by comparing
+  Each device is represented by a subclass of ``Device`` (e.g. ``CpuDevice``,
+  ``GpuDevice``). The length of the returned list is equal to
+  ``device_count(backend)``. Local devices can be identified by comparing
   ``Device.host_id`` to ``host_id()``.
 
+  If ``backend`` is ``None``, returns all the devices from the default backend.
```
Comment (gnecula) [inline jax/lib/xla_bridge.py:195]: I think it is useful to explain that backends have a priority

File: jax/lib/xla_bridge.py [hunk 3]
```diff
@@ -179,19 +179,21 @@ def device_count(backend=None):
...
 
-def devices(backend=None):
-  """Returns a list of all devices.
+def devices(backend: str = None):
+  """Returns a list of all devices for a given backend.
 
-  Each device is represented by a subclass of Device (e.g. CpuDevice,
-  GpuDevice). The length of the returned list is equal to
-  ``device_count()``. Local devices can be identified by comparing
+  Each device is represented by a subclass of ``Device`` (e.g. ``CpuDevice``,
```
Comment (gnecula) [inline jax/lib/xla_bridge.py:190]: A very important concept that is not described here is that of the default device: first device on the default backend. Or is it, first local_device on the default backend?

**Review by gnecula — APPROVED** [review]

File: jax/lib/xla_bridge.py (hunk 3 above)
Comment (skye) [reply jax/lib/xla_bridge.py:190]: It should be the first local device, but I'm not 100% sure that's true on TPU pods (we use the default XLA device assignment, which in theory can do whatever it wants). We should probably explicitly set a device in jax in the default single-device case, so I'm gonna leave this out for now.

File: jax/lib/xla_bridge.py (hunk 2 above)
Comment (skye) [reply jax/lib/xla_bridge.py:195]: Done.

File: jax/lib/xla_bridge.py (hunk 1 above)
Comment (skye) [reply jax/lib/xla_bridge.py:210]: I just tried it. A bad backend raises an error, but a bad host_id currently returns an empty list. I made it raise an error and added a small test.

**Pull request merged**